sdk/storage/Azure.Storage.Common/src/Shared/LazyLoadingReadOnlyStream.cs

```diff
@@ -271,7 +263,11 @@ private async Task<int> DownloadInternal(bool async, CancellationToken cancellat
             // Set _requestConditions If-Match if we are not allowing the blob to be modified.
             if (!_allowBlobModifications)
             {
-                _requestConditions = _createRequestConditionsFunc(response.GetRawResponse().Headers.ETag);
+                // TODO (kasobol-msft) find better exception.
+                if (_expectedETag.Value != response.GetRawResponse().Headers.ETag)
```

Comment:
Why are we comparing this/throwing on the client-side when If-Match should cause a 412?

Reply:
In case of Shares the `ShareFileRequestConditions` contains only `LeaseId`. Shares don't support `If-*` conditions at all. 
Which makes the [allowModifications](https://docs.microsoft.com/en-us/dotnet/api/azure.storage.files.shares.models.sharefileopenreadoptions.-ctor?view=azure-dotnet) option there ineffective.
This is intended to close this gap.

https://github.com/Azure/azure-sdk-for-net/blob/e88e7d4a4973be2c667c2f57c8d5d3bd30ccb6a0/sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs#L2261

Reply:
Please add a comment that this is specific to Shares.  I could see being more explicit and taking in a parameter to explicitly declare you want client-side validation here.  I'd throw an `InvalidOperationException` since it's not tied to a request.  But I'm kind of curious what we think customers would do with this - do they need any additional information to recover gracefully?

Reply:
Initially I was thinking about forging 412 error, but it isn't great idea - they'll see 2xx if they enable sdk logging.
Does dotnet have something like [ConcurrentModificationException](https://docs.oracle.com/javase/7/docs/api/java/util/ConcurrentModificationException.html)?
They have to know that exception being served is due to file being modified concurrently. If not for retry (or some other recovery strategy) then just for diagnostic purposes - so they know they have a race in the system.
InvalidOperationException with clear message solves the later but might be inconvenient to handle programatically if they want to do something sophisticated (i.e. take a lease as fallback strategy?).

Reply:
We don't have a great one.  https://docs.microsoft.com/en-us/dotnet/api/system.data.linq.changeconflictexception?view=netframework-4.8&viewFallbackFrom=net-5.0 exists but it's obscure.  You can scroll through the derived classes section of https://docs.microsoft.com/en-us/dotnet/api/system.exception?view=net-5.0 and https://docs.microsoft.com/en-us/dotnet/api/system.systemexception?view=net-5.0 to look for other things that might work.

This might be worth a custom exception type then to include enough context like the URI of the thing being read, the old etag/new etag, the current location in the stream, etc.

Reply:
How about https://docs.microsoft.com/en-us/dotnet/api/system.io.invaliddataexception?view=net-5.0 with `Data` filled with contextual information? Otherwise custom exception.

Reply:
Unfortunately nobody ever finds anything in `Data`.  I'd recommend a custom exception if we think people will actually be able to recover.